### ISSUE TITLE
CoffeeScript and Philosophy

### ISSUE THREAD

**Issue by maedi** [tower-archive/tower#216]
Just a few thoughts in reply to recent IRC conversations:
### CoffeeScript

**viatropos:** "one of the next things is to make a pure JS api, probably will be the default as well [...] as much as I like coffeescript it doesn't feel it is on the upswing anymore. a larger percent of people understand javascript too, so having that the default will lower the barrier"

**e_dub:** "its sad to me, because coffeescript is actually my favorite language. And I really like the way models are currently in tower. but i do agree with you. there's been a strong backlash against coffeescript by some people , which really doesnt make sense to me. But as long as "I" can still use coffeescript, that's cool."

**luckysmack:** "to be honest, i started looking at frameworks a few weeks ago. and if tower was in js i would have already chosen it. but after seeing other frameworks and not liking how some of them do things, and after a couple people recommended it, i started looking again earlier today, which is when I asked."

Personally I love CoffeeScript and wouldn't have moved from Ruby to Node without it. I think Tower should keep 'simplicity' as its main aim and not try to appeal to the hardcore node community. Derby and Meteor do this, they are verbose frameworks with a lack of structure/convention which leave most of us feeling confused. Tower is attracting a much larger audience of the common coder who are seeking elegance and simplicity. Many will come from Ruby and will understand it instantly if it's in CoffeeScript.

Tower will become incredibly popular because it makes sense, both in structure and in syntax, not because it follows the Node community.

I can see the new generations learning CS instead of JS, much like we all learn jQuery over JS now for the DOM. Could we please keep CoffeeScript as the default? And have JS as the optional? Apologies but I feel this one's important.
### Model/View Bindings

**viatropos:** "for the model/view binding, we just merged that functionality to master. tower's client stuff is built around ember. [...] the rails/activerecord query api, like App.User.where(email: 'x').page(2)... creates a "cursor", and you can register that cursor with both client-side and server-side controllers (right now using the `@scope App.User.where...` class method on the controller. That then sets up the data-syncing functionality, so whenever a record is created/updated/deleted matching the criteria in the cursor, it will be added to the cursor. You can then bind those cursors to views/lists/etc and everything will automatically update"

Great work with integrating Ember with Tower. I'm really looking forward to figuring out how this works!
### File structure

+1, this new directory structure looks good:
https://gist.github.com/1398757#gistcomment-395680
### MVC

**e_dub:** "also, definitely with you on keeping with mvc. I get the excitement around some of the newer conceptual models, but part of what makes tower so great, in my opinion, is that it feels so much like rails without compromising on staying true to node"

As you guys know the Node community is trying to move away from traditional MVC. But as viatropos mentioned in a comment in the above gist, none of the other MV*'s seem to work on a practical level. I totally agree but I'm just curious about whether MVC impacts the real-time features? I've yet to make the mental leap between Tower and Ember so bare with me...

Overall really happy with Tower as a framework, well thought out and a solid foundation. After looking at everything else I believe Tower is the one. But please be the Rails of the Node world and use beautiful code ;)

**Comment by autoferrit** [comment]
In all honesty, I was never a bug fan of coffee script. But reading through the docs page, and seeing how it also translates into html (reminds me of haml), at that point, i thought 'whoah that is cool and super easy to read' . while looking through the examples im starting to understand who coffee compares to vanilla js, some areas im still not sure about. But, if your main love is CS and you want the core to be expressed in that way, then keep it. Offer vanilla js as an alternative. it doesnt mean you cant have both options available.

tower new --vanilla myApp (for straight js)

which would set a variable somewhere in the generated app to use non coffee for the project. also keep in mind you could configure coffeekup separately. because after seeing it I could still consider using that for html files. it is kinda cool. since coffee is supposed to compile into js so reliably, I think you could potentially support both, which means you could cater to both crowds. When you have examples, you could set them up in a way to show coffee and have a tab at the top of a code snippet to show either the coffee or plain js version of the code. its the same thing microsoft does for their .net api. you have the same api and choose which language to view the example in.

I think being able to support both could get the best of both worlds audience wise. and leaving the default as coffee would still leave you with the warm fuzzy feeling you get. and when others start using it with js, you can still get bug reports and if they are interested some may even help with keeping the api for plain js.

basically if you love CS as much as you say, I would keep it as js, offer the plain api, and have both options. I think it would get more people willing to try coffee (have a script to convert project files to coffee), and could get more people to use it. also helping get more peeps to like coffee in the mornings.

**Comment by lancejpollard** [comment]
@Maedi great points. This is definitely something to think a lot more about.

**Comment by ajb** [comment]
Just want to chime in here, hope I'm not out of place since I've only recently found out about this project and am not a contributor (yet).

From my outside standpoint, Tower looks _amazing_ and I can't wait for it to be production-ready. So I was a little disappointed when I saw that development time was going to be spent on converting the already-written source to javascript, especially since there's nothing keeping the end-user from using javascript in their project, even if Tower is written in Coffee.

**Comment by maedi** [comment]
I haven't contributed any code either so apologies as well, but I agree with you @adamjacobbecker

**Comment by mjcloutier** [comment]
I agree with @maedi and @adamjacobbecker Tower.js looks absolutely beautiful with coffeescript, hopefully it will remain the default. Looking forward to studying/developing with it soon!

**Comment by lancejpollard** [comment]
@adamjacobbecker The main thinking behind converting the CS to JS are to streamline development and to make tower's source accessible to the plain JS developers (of which there's potentially ~5x as many from looking at the GitHub language breakdown). We can convert it by just using the generated JS. I might make a few "style" changes to the generated JS, but other than that we wouldn't be spending much time on it.

@viotech @maedi and @adamjacobbecker I agree with you all that the CS version looks super nice. However we're facing a couple of challenges with that API (and CoffeeScript in general) that's getting in the way of development. Mainly:
1. We have to spend a lot of time debugging CoffeeScript when new users start trying to use the app.
2. The CS extensions we've created ([./coffee-inheritance.js](https://github.com/viatropos/tower/blob/master/coffee-inheritance.js)) aren't compatible with CS 1.4.0 and need to be updated.
3. I haven't figured out a good way to have the "[executable class bodies](http://yehudakatz.com/2009/06/04/the-importance-of-executable-class-bodies/)" that CS allows work well with the `Ember.Object.extend` API (which is how most JS APIs are written).
4. Since the client is more complicated than the server in some cases (mainly, routes), the `extend({})` syntax is _better_ than the DSL in some cases (and the number of cases it's better in seems to be growing).
5. Having to compile the CS is a turnoff to a large amount of potential contributors, and it also slows down TDD - to develop new features for the client we need to compile files individually and use something like Require.js, but we're not there yet so we have to wait for the entire project to compile for every change! (maybe this is just me though, definitely open to suggestions).

The first issue takes a way a lot of time/energy/thought from actual development. If we can get rid of this issue somehow (any ideas?), then that would be a huge plus. It pretty much means solving the other 4 challenges.

For the second, there is an issue with the current implementation when you minify the generated JS for production. This would take a few days or a week to fix since I am no expert in language grammars.

The third issue is the biggest. Mainly, there needs to be a clear way to append attributes/methods to the prototype (I think it's accessible through `Ember.Object.proto()`), but should be added as a mixin. Take the model definition now:

``` coffeescript
class App.Post extends Tower.Model
  @field 'title'
  @field 'text'
```

Currently that is creating 2 mixins, and for every new field it will create a new mixin. That's because `@field` is essentially just doing `@reopen(title: new Tower.Model.Field)` for each field. I tried optimizing this by hacking into the _last_ mixin, but it's very brittle and I don't know if there are ways this could cause unpredictable errors. You can see how I did this here:
- [packages/tower-model/shared/attributes.coffee#L45-47](https://github.com/viatropos/tower/blob/2376f81f1e37897f073db300a4d9d78b96a29572/packages/tower-model/shared/attributes.coffee#L45-47)
- The hack: [packages/tower-model/shared/attribute.coffee#L153-158](https://github.com/viatropos/tower/blob/master/packages/tower-model/shared/attribute.coffee#L153-158)

So currently that hack works, for the simple case of defining fields on initialize. But the problem is if any of these:
1. If you define fields _at run time_ (after Ember is initialized) we may need to "invalidate" the class' mixins. I'm not sure about this, but to find out we need to write some tests (or deeply understand Ember's source) so we know we _don't_ have to invalidate them. This takes time, it could take a few days to resolve this.
2. There's issues doing it this way with binding and computed properties that I've run into that were hard to debug (can't recall exactly), so I never dug in further. But it had something to do with auto-updating associations when a record in the system changed I think.

So at this point you can see, a lot of the time I've been spending (or thought's I've had to think through) are how to get CoffeeScript APIs to play nice with JS APIs - or, specifically, Ember. One thought is that changing the source code to JS would eliminate this entire field from thought, so I can start focusing on core functionality.

The final issue is that things like the routes look better using the `Ember.Object.extend` API vs. the Rails-like DSL:
- https://github.com/viatropos/tower/pull/299

So at first we started with these APIs using the Rails-like DSL:
- routes
- models (fields/associations/validations/callbacks)
- controllers (params/scopes/callbacks)

But now that we're thinking of doing the router with the JS/Object API, we're down to just models/controllers using the DSL. But because we're also thinking about changing the functionality of the controllers (see that routes pull request above), and moving the "actions" to the router itself, then we don't need callbacks in the controllers anymore, and we also can get rid of the class methods `param` and `scope` so you just create a new controller for the scopes, and params are defined in the router (also detailed in the routes pull request).

This means the controller doesn't have any "executable class bodies" anymore; it doesn't need them.

So now it's just the models.

I agree the field/association definitions look nicer in CS. But this is the only part of Tower that uses it. And in JS, it is very verbose. Then you add on top of these 2 facts the items from the list above (time spent debugging CS for newcomers, our custom CS extensions not up-to-date, and the non-ideal-ness / performance-loss of creating a new mixin for every field definition), it seems like a waste of time/effort to keep that CS API.

I personally don't want to work on that atm, I am working on some projects applying Tower and find it is easier just writing the code in JS even if it's not as pretty. However, if the CS API was all working well and we didn't have to spend time regularly making sure it still works, and it was just as optimized as a JS equivalent, then I would probably use it. If any of you want to take the lead on that I will be there to help whenever you need it. Let me know.

Thoughts?

**Comment by edubkendo** [comment]
Ah damn, so the coffeescript API is going to quit being supported entirely?

**Comment by thehydroimpulse** [comment]
I'm guessing, if we could fix these issues with coffee-script, in particular, https://github.com/emberjs/ember.js/issues/599 , I think could lead to a different coffee-script api then it's Javascript counter part. This means, you could use the simplified, or Ember-like, Javascript API, or use the traditional coffee-script classes like it's done atm. 

The main problem, though, is the client-side development process with Tower. Atm it's pretty complex and requires a ton of overhead to build a single file. Coffee-script isn't the biggest culprit, but it's a small factor in the process.

**Comment by edubkendo** [comment]
So, it seems like the issue (ignoring a couple of issues which are getting kind of smashed together here, but really need to be talked about entirely seperately) is in trying to force coffee-script to generate code that isn't really the Javascript the compiler was designed to generate in an attempt to make the original Tower API fit 1:1 with Ember's API. So, I think the solution, for those of us who want to keep using coffee-script, is to figure out what needs to change in order to just use the plain-jane coffee-script compiler. I've written plenty of Ember code in coffee-script, and plenty of server-side node code too, using just ordinary coffee-script, and imo it's still far preferable to writing either in javascript, even if it means that it winds up looking a little bit different from the way Tower code is written currently. That API is kind of in a state of flux anyways. 

As far as the part about waiting while it compiles to javascript, I think the solution to that is pretty easy too, DON'T compile to javascript. When I write coffee-script for other projects I never do, I just require coffee-script at my app's entry point, and use it that way. it never makes debugging an issue, you just skip looking at line numbers, and instead you look at the function and object names. Besides, source maps are working in the coffee-script redux project, which is supposed to be a drop in replacement for the current compiler, so it gives you totally accurate coffee-script line numbers, both in the browser and in node.

I kind of don't know where to start with all this, and I'm not sure if I want to take it on, I need to give it some thought. I guess a starting point would be to figure out exactly where Ember and coffee-script are currently butting heads, so anyways, there's that. I'll comment on some of the other stuff at issue later. Any thoughts guys?

**Comment by lancejpollard** [comment]
@edubkendo interesting notes about coffeescript redux / source maps, will have to check that out at some point.

I think I've been confusing terms when talking about CS/JS with Tower. The main thing I'm concerned with is that the "executable class bodies" part of the CoffeeScript API is not working very nicely with Ember. And when I've been saying "JS API" I don't mean writing the code in JavaScript, but rather, using the "object" API (or whatever we want to call it), which can be written in CS too:

``` coffeescript
App.Post = Tower.Model.extend
  title: Tower.Model.field('string')
  body: Tower.Model.field('string')
```

> Note, the API in that code is not the final API, just ideas atm.

What I meant by "moving from CS to JS" is basically removing the _executable class body API_. So, the default for a generated tower app could still be CS, I'm cool with that. But the main issue is, unless we can get the executable class body API in CS to be just as good/error-free/optimized as the "object API" (above), then when you generate a tower app it should generate "object API" code like above rather than the current executable class body api:

``` coffeescript
class App.Post extends Tower.Model
  @field 'title'
  @field 'text'
```

Getting rid of the executable class body API would make it so we can use plain-jane coffee-script. The main thing we need to decide on, then, is an "object API", and if everyone's cool with either 1) getting rid of the "executable class body API" in favor of the "object API", or 2) having both APIs, and defaulting to one of them.

**Comment by lancejpollard** [comment]
Really, we could do the executable class bodies API as well, it would just take a decent amount more time to implement a more generic extension module for coffeescript, where we can more easily make changes. This would also set us up to hook into other x-2-js languages.

**Comment by edubkendo** [comment]
_breathes a HUGE sigh of relief_

fuck me backwards, lol. That's a huge difference.

I'm fine with changing the API to use Ember.extends instead of CS classes in the places where trying to use classes makes it a problem. I rarely use coffee-script's classes in my own code anyways. Usually just for meta-programming stuff, which would still be do-able in this situation. 

As far as the executable class bodies go, I don't know a lot yet about the problems or their solutions, because I haven't really looked at it deeply before now, but [CoffeeScriptRedux](https://github.com/michaelficarra/CoffeeScriptRedux) has extensibility hooks that might make it feasible to easily address these issues. Check out: [Hacking the CoffeeScript Redux Compiler](http://codebrief.com/2012/08/hacking-the-coffeescript-compiler/) and [EmberScript](https://github.com/ghempton/ember-script) for some examples.  Finally, here's a bit of info on [CoffeeScript Source Maps](http://ryanflorence.com/2012/coffeescript-source-maps/). Just glancing at the list of his current issues, it looks like he's in the home stretch now and just cleaning up the bugs, so it's probably time we take a serious look at this.

Anyways, really glad that got cleared up as I was at wit's end. So, to be clear, `tower new myVeryAwesomeApp` will still generate .coffee files, or at least something like `tower new myVeryAwesomeApps --use-coffee` will? They just will have a more Ember-like and slightly less Rails-like API to them? This isn't a big deal for me. I really like the current API for models, but I was the one pushing hardest for the Ember-style router, or at least for keeping access to the Ember style router, so I am ok with re-working the Model API to something that doesnt cause problems. I'm definitely on the same page with you then, I think.

**Comment by edubkendo** [comment]
See this comment, also, btw: http://codebrief.com/2012/08/hacking-the-coffeescript-compiler/#comment-621296454

**Comment by lancejpollard** [comment]
Not going to change the source or tests to JavaScript until after we try using [CoffeeScriptRedux](http://www.kickstarter.com/projects/michaelficarra/make-a-better-coffeescript-compiler/posts) (thanks again for pointing out @edubkendo). If the compiler is as extensible as it seems to be (it doesn't seem to be complete yet https://github.com/michaelficarra/CoffeeScriptRedux/issues/101), then this could open up a lot of doors. Also, I want to try the new grunt watcher and see if we get significant performance improvements.

If we can customize the coffeescript compiler, and the watcher doesn't use up a bunch of memory, then it might mean keeping the source in coffeescript could be easier than javascript. Right now its kind of painful though, after a while.

**Comment by maedi** [comment]
After shedding a tear for the elegance of 'self executing class bodies' I too would be happy with a more Ember like CoffeeScript API. Making Tower behave like Ember is a good move. EmberScript is a nice idea too.

Hopefully the future compiler will allow the `@field` syntax and work with Ember? Ultimately you want the same syntax everywhere.

**Comment by lancejpollard** [comment]
We are moving to JavaScript. It's so much more powerful, and I don't think CoffeeScript will be around much longer with ES6. The main thing is, plain JS is much much faster than CoffeeScript. It will lower the barrier to entry as well.

**Comment by thehydroimpulse** [comment]
:+1:
